### PR TITLE
Turn off viewcode extension for epub by default

### PR DESCRIPTION
**Pull request by shibukawa**
epub checker doesn't prefer pages outside toctree. This fix improves epub checker scores.

Before:
- 65 warnings, 1247 errors

After:
- 65 warnings, 867 errors

I added opt-in option to enable viewcode extension for epub building.

### CODE REVIEW

**Comment by tk0miya** [comment]
Hmm. Is there any reason to add a option to enable viewcode?
As a quick look, all viewcode do not work in ePub format to me. At least, ibooks could not trace the source code links by viewcode.

I agree with disabling the extension. But we have to discuss to support enabling viewcode on epub, I think.

**Comment by shibukawa** [comment]
I added new option to enable keeping backward compatibility, but turning-off completely is good/simple idea if others can accept it.

**Comment by tk0miya** [comment]
In other words, the `viewcode_enable_epub` option says viewcode works fine with EPUB if enabled. Is this true?
If not, I wonder why we provides the option.

Of course it would be nice if viewcode supports EPUB output. But we have to describe what the option does and the restriction of it.

**Comment by shibukawa** [comment]
I added warning section to viewcode's document.

**Comment by tk0miya** [comment]
It seems viewcode does not work with other builders. I confirmed with latex, text and singlehtml builders.
As a quick look, it uses `html-collect-pages` handler. So the extension only works with HTML builder and its descendants.
I don't confirm other HTML builder families yet. We need to check them.

IMO, it would be nice if we admit that the viewcode extension only works with bare HTML builder, and clearly say that in docs.

**Comment by tk0miya** [comment]
In my short investigation:
- applehelp
  - Fine
- devhelp
  - I can't confirm this (because of environments)
- epub
  - Generated, but readers could not trace hyperlinks
- html/StandaloneHTMLBuilder
  - Of course, it works
- html/DirectoryHTMLBuilder
  - Generated, but links are broken
- html/SingleFileHTMLBuilder
  - Files are not generated, so links are broken
- html/PickleHTMLBuilder
  - I don't know how the builder works...
- html/JSONHTMLBuilder
  - I don't know how the builder works... (it seems works)
- htmlhelp
  - I can't confirm this (because of environments)
- qthelp
  - I can't confirm this (because of environments)

**Comment by shibukawa** [comment]
- qthelp
  - Works
- devhelp
  - Works

**Comment by tk0miya** [comment]
Thank you for investigation.

It looks the extension does not support `dirhtml` and `signlehtml` too.

About `dirhtml`, it seems a bug of the builder. So we should file this as new issue.

Next, about `singlehtml`, no documents are generated by viewcode. So it is not supported at this moment. So it would be nice to disable it now.

Last, the main theme of this PR, about `epub`, I'm okay to provide the `viewcode_enable_epub` option. I know it generates the documents which is out of the EPUB specs by default. But users can link the them to the toctree manually, and it will make valid EPUB docs.

File: CHANGES [hunk 1]
```diff
@@ -31,13 +31,14 @@ Incompatible changes
...
+  as a docclass by default if the ``language`` is ``ja``.
 * latex: To provide good default settings to Japanese docs, Sphinx uses
   ``jreport`` and ``jsbooks`` as a docclass by default if the ``language`` is
   ``ja``.
-* `sphinx-quickstart` now allows a project version is empty
+* ``sphinx-quickstart`` now allows a project version is empty
 * Fix :download: role on epub/qthelp builder. They ignore the role because they don't support it.
-
+* ``sphinx.ext.viewcode`` doesn't work on epub building by default. ``viewcode_enable_epub`` option
+  is added to enable this extension.
```
Comment (tk0miya) [inline CHANGES]: Could you add incompatible changes entry for singlehtml?

**Comment by tk0miya** [comment]
LGTM with nits.
I will merge this after fixed.

**Comment by tk0miya** [comment]
Merged. Thank you for great work!

**Pull request merged**